### PR TITLE
[LoRA] Adds example on text2image fine-tuning with LoRA

### PR DESCRIPTION
**Pull request by sayakpaul**
Very much copied from @patrickvonplaten's awesome work with https://github.com/huggingface/diffusers/pull/1884

Things seem to be working on both T4 and V100. 

My command:

```bash
export MODEL_NAME="CompVis/stable-diffusion-v1-4"
export DATASET_NAME="lambdalabs/pokemon-blip-captions"

accelerate launch --mixed_precision="fp16" \
  train_text_to_image_lora.py \
  --pretrained_model_name_or_path=$MODEL_NAME \
  --dataset_name=$DATASET_NAME --caption_column="text" \
  --resolution=512 --random_flip \
  --train_batch_size=1 \
  --num_train_epochs=100 --checkpointing_steps=5000 \
  --learning_rate=1e-04 --lr_scheduler="constant" --lr_warmup_steps=0 \
  --seed=42 \
  --enable_xformers_memory_efficient_attention \
  --validation_prompt="cute Sundar Pichai creature" --report_to="wandb" \
  --output_dir="sd-model-finetuned-lora-t4" \
  --push_to_hub && sudo shutdown now
```

The final weights will be pushed to https://huggingface.co/sayakpaul/sd-model-finetuned-lora-v100 and an experimentation run is available here: https://wandb.ai/sayakpaul/text2image-fine-tune/runs/3iiol807 (currently running). Once these are done, I will update the appropriate sections in the README.

### CODE REVIEW

File: examples/text_to_image/README.md [hunk 1]
```diff
@@ -110,7 +110,80 @@ image = pipe(prompt="yoda").images[0]
...
+With LoRA, it's possible to fine-tune Stable Diffusion on a custom image-caption pair dataset
+on consumer GPUs like Tesla T4, Tesla V100.
+
+### Training
+
+First, you need to set-up your dreambooth training example as is explained in the [installation section](#installing-the-dependencies). Make sure to set the `MODEL_NAME` and `DATASET_NAME` environment variables. Here, we will use [Stable Diffusion v1-4](https://hf.co/CompVis/stable-diffusion-v1-4) and the [Pokemons dataset](https://hf.colambdalabs/pokemon-blip-captions).  
+
+**___Note: Change the `resolution` to 768 if you are using the [stable-diffusion-2](https://huggingface.co/stabilityai/stable-diffusion-2) 768x768 model.___**
+
+**___Note: It is quite useful to monitor the training progress by regularly generating sample images during training. [Weights and Biases](https://docs.wandb.ai/quickstart) is a nice solution to easily see generating images during training. All you need to do is to run `pip install wandb` before training to automatically log images.___**
```
Comment (patil-suraj) [inline examples/text_to_image/README.md:134]:
```suggestion

First, you need to set up your dreambooth training example as is explained in the [installation section](#installing-the-dependencies). Make sure to set the `MODEL_NAME` and `DATASET_NAME` environment variables. Here, we will use [Stable Diffusion v1-4](https://hf.co/CompVis/stable-diffusion-v1-4) and the [Pokemons dataset](https://hf.colambdalabs/pokemon-blip-captions).  

**___Note: Change the `resolution` to 768 if you are using the [stable-diffusion-2](https://huggingface.co/stabilityai/stable-diffusion-2) 768x768 model.___**

**___Note: It is quite useful to monitor the training progress by regularly generating sample images during training. [Weights and Biases](https://docs.wandb.ai/quickstart) is a nice solution to easily see generating images during training. All you need to do is to run `pip install wandb` before training to automatically log images.___**
```

File: examples/text_to_image/README.md [hunk 2]
```diff
@@ -110,7 +110,80 @@ image = pipe(prompt="yoda").images[0]
...
+  --seed=42 \
+  --output_dir="sd-pokemon-model-lora" \
+  --save_sample_prompt="cute dragon creature" --report_to="wandb"
+```
+
+The above command will also run inference as fine-tuning progresses and will log the results to Weights and Biases.
+
+**___Note: When using LoRA we can use a much higher learning rate compared to non-LoRA fine-tuning. Here we use *1e-4* instead of the usual *1e-5*.___**
+
+The final LoRA embedding weights have been uploaded to [TODO](TODO). **___Note: [The final weights](TODO) are only 3 MB in size which is orders of magnitudes smaller than the original model.**
```
Comment (patil-suraj) [inline examples/text_to_image/README.md:167]:
```suggestion
The above command will also run inference as fine-tuning progresses and log the results to Weights and Biases.

**___Note: When using LoRA we can use a much higher learning rate compared to non-LoRA fine-tuning. Here we use *1e-4* instead of the usual *1e-5*.___**

The final LoRA embedding weights have been uploaded to [TODO](TODO). **___Note: [The final weights](TODO) are only 3 MB in size, which is orders of magnitudes smaller than the original model.**
```

File: examples/text_to_image/train_text_to_image_lora.py [hunk 3]
```diff
@@ -0,0 +1,824 @@
+# coding=utf-8
+# Copyright 2023 The HuggingFace Inc. team. All rights reserved.
```
Comment (patil-suraj) [inline examples/text_to_image/train_text_to_image_lora.py:2]: (nit)

maybe also mention the original lora repo in copyright.

File: examples/text_to_image/train_text_to_image_lora.py [hunk 4]
```diff
@@ -0,0 +1,824 @@
...
+    noise_scheduler = DDPMScheduler.from_pretrained(args.pretrained_model_name_or_path, subfolder="scheduler")
+    tokenizer = CLIPTokenizer.from_pretrained(
+        args.pretrained_model_name_or_path, subfolder="tokenizer", revision=args.revision
+    )
+    text_encoder = CLIPTextModel.from_pretrained(
+        args.pretrained_model_name_or_path, subfolder="text_encoder", revision=args.revision
+    )
+    vae = AutoencoderKL.from_pretrained(args.pretrained_model_name_or_path, subfolder="vae", revision=args.revision)
+    unet = UNet2DConditionModel.from_pretrained(
+        args.pretrained_model_name_or_path, subfolder="unet", revision=args.non_ema_revision
```
Comment (patil-suraj) [inline examples/text_to_image/train_text_to_image_lora.py:420]: we are not using ema in this script
```suggestion
        args.pretrained_model_name_or_path, subfolder="unet", revision=args.revision
```

File: examples/text_to_image/train_text_to_image_lora.py [hunk 5]
```diff
@@ -0,0 +1,824 @@
...
+    if env_local_rank != -1 and env_local_rank != args.local_rank:
+        args.local_rank = env_local_rank
+
+    # Sanity checks
+    if args.dataset_name is None and args.train_data_dir is None:
+        raise ValueError("Need either a dataset name or a training folder.")
+
+    # default to using the same revision for the non-ema model if not specified
+    if args.non_ema_revision is None:
+        args.non_ema_revision = args.revision
```
Comment (patil-suraj) [inline examples/text_to_image/train_text_to_image_lora.py:337]: same comment as above
```suggestion
```

File: examples/text_to_image/README.md (hunk 2 above)
Comment (patrickvonplaten) [inline examples/text_to_image/README.md:167]: Should we maybe add final links here before merging? :-) Like we do one final run now and then we have all the links here?

File: examples/text_to_image/README.md [hunk 6]
```diff
@@ -110,7 +110,80 @@ image = pipe(prompt="yoda").images[0]
...
+### Inference
+
+Once you have trained a model using above command, the inference can be done simply using the `StableDiffusionPipeline` after loading the trained LoRA weights.  You 
+need to pass the `output_dir` for loading the LoRA weights which, in this case, is `sd-pokemon-model-lora`.
+
+```python
+from diffusers import StableDiffusionPipeline
+import torch
+
+model_path = "path_to_saved_model"
```
Comment (patrickvonplaten) [inline examples/text_to_image/README.md:178]: I'd replace this `"path_to_saved_model"` with the equivalent of: https://huggingface.co/patrickvonplaten/lora_dreambooth_dog_example -> it's always nice when code examples just work out of the box

File: examples/text_to_image/train_text_to_image_lora.py [hunk 7]
```diff
@@ -0,0 +1,824 @@
...
+from transformers import CLIPTextModel, CLIPTokenizer
+
+
+# Will error if the minimal version of diffusers is not installed. Remove at your own risks.
+check_min_version("0.12.0.dev0")
+
+logger = get_logger(__name__, log_level="INFO")
+
+
+class LoraLayers(torch.nn.Module):
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image_lora.py:55]:
```suggestion
```

We can remove the whole class here and just import it from as done here: https://github.com/huggingface/diffusers/blob/7c82a16fc14840429566aec40eb9e65aa57005fd/examples/dreambooth/train_dreambooth_lora.py#L43

File: examples/text_to_image/train_text_to_image_lora.py [hunk 8]
```diff
@@ -0,0 +1,824 @@
...
+    parser.add_argument(
+        "--num_validation_images",
+        type=int,
+        default=4,
+        help="Number of images that should be generated during validation with `validation_prompt`.",
+    )
+    parser.add_argument(
+        "--validation_epochs",
+        type=int,
+        default=10,
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image_lora.py:148]: For fine-tuning I'd say we set the default a bit lower no? In dreambooth one usually has only 5 images which means that 10 epochs == 50 samples which is only 25 steps if batch size is 2. 

Here we can maybe set this just to 1 as a default? 
```suggestion
        default=1,
```

File: examples/text_to_image/train_text_to_image_lora.py [hunk 9]
```diff
@@ -0,0 +1,824 @@
...
+    if token is None:
+        token = HfFolder.get_token()
+    if organization is None:
+        username = whoami(token)["name"]
+        return f"{username}/{model_id}"
+    else:
+        return f"{organization}/{model_id}"
+
+
+dataset_name_mapping = {
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image_lora.py:352]: Very nice! 
Just a nit since it's IMO a constant:
```suggestion
DATASET_NAME_MAPPING = {
```

**Review by patil-suraj — APPROVED** [review]
Thanks a lot for working on this! Just left some nits. And we should also add the model card creation function in this script. cf #2032

File: examples/text_to_image/train_text_to_image_lora.py [hunk 10]
```diff
@@ -0,0 +1,824 @@
...
+            hidden_size = unet.config.block_out_channels[block_id]
+
+        lora_attn_procs[name] = LoRACrossAttnProcessor(
+            hidden_size=hidden_size, cross_attention_dim=cross_attention_dim
+        )
+
+    unet.set_attn_processor(lora_attn_procs)
+    lora_layers = AttnProcsLayers(unet.attn_processors)
+
+    state_dict = lora_layers.state_dict()
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image_lora.py:475]:
```suggestion
```

File: examples/text_to_image/train_text_to_image_lora.py [hunk 11]
```diff
@@ -0,0 +1,824 @@
...
+    lora_layers.load_state_dict(state_dict)
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image_lora.py:476]:
```suggestion
```

Comment (patrickvonplaten) [reply examples/text_to_image/train_text_to_image_lora.py:476]: I just added those two lines for debugging and removed them later :-)

File: examples/text_to_image/train_text_to_image_lora.py [hunk 12]
```diff
@@ -0,0 +1,824 @@
...
+                torch_dtype=weight_dtype,
+            )
+            pipeline = pipeline.to(accelerator.device)
+            pipeline.set_progress_bar_config(disable=True)
+
+            # run inference
+            generator = torch.Generator(device=accelerator.device).manual_seed(args.seed)
+            images = []
+            for _ in range(args.num_validation_images):
+                images.append(pipeline(args.validation_prompt, num_inference_steps=30, generator=generator).images[0])
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image_lora.py:769]: nice! running in a loop is more memory efficient here :-) We should maybe change this in the dreambooth lora example as well

File: examples/text_to_image/train_text_to_image_lora.py [hunk 13]
```diff
@@ -0,0 +1,824 @@
...
+
+    for tracker in accelerator.trackers:
+        if tracker.name == "wandb":
+            tracker.log(
+                {
+                    "test": [
+                        wandb.Image(image, caption=f"{i}: {args.validation_prompt}") for i, image in enumerate(images)
+                    ]
+                }
+            )
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image_lora.py:818]: Can we here also add the "model creation" part: https://github.com/huggingface/diffusers/blob/7c82a16fc14840429566aec40eb9e65aa57005fd/examples/dreambooth/train_dreambooth_lora.py#L61 

Feel free to just copy paste :-)

**Review by patrickvonplaten — COMMENTED** [review]
Super nice! Great that you made it work so quickly. Think only a couple of things are left to do :-)

File: examples/text_to_image/train_text_to_image_lora.py (hunk 12 above)
Comment (sayakpaul) [reply examples/text_to_image/train_text_to_image_lora.py:769]: Yes, that is why I had do it this way rather than one-shot. The latter was leading to OOM.

File: examples/text_to_image/README.md (hunk 2 above)
Comment (sayakpaul) [reply examples/text_to_image/README.md:167]: Yes, that is what my plan is :)

File: examples/text_to_image/README.md (hunk 6 above)
Comment (sayakpaul) [reply examples/text_to_image/README.md:178]: Yes, will do!

File: examples/text_to_image/train_text_to_image_lora.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/text_to_image/train_text_to_image_lora.py:769]: Yes.

File: examples/text_to_image/train_text_to_image_lora.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (sayakpaul) [reply examples/text_to_image/train_text_to_image_lora.py:818]: Wrapped it under: `accelerator.is_main_process`.

File: examples/text_to_image/train_text_to_image_lora.py (hunk 3 above)
Comment (sayakpaul) [reply examples/text_to_image/train_text_to_image_lora.py:2]: [This](https://github.com/microsoft/LoRA) or [this](https://github.com/cloneofsimo/lora)?

File: examples/text_to_image/README.md (hunk 2 above)
Comment (sayakpaul) [reply examples/text_to_image/README.md]: Done.

File: examples/text_to_image/README.md (hunk 6 above)
Comment (sayakpaul) [reply examples/text_to_image/README.md]: Done.

**Comment by sayakpaul** [comment]
@patil-suraj @patrickvonplaten I think we're good to go :)

File: examples/text_to_image/README.md [hunk 14]
```diff
@@ -110,7 +110,82 @@ image = pipe(prompt="yoda").images[0]
...
+  --seed=42 \
+  --output_dir="sd-pokemon-model-lora" \
+  --save_sample_prompt="cute dragon creature" --report_to="wandb"
+```
+
+The above command will also run inference as fine-tuning progresses and log the results to Weights and Biases.
+
+**___Note: When using LoRA we can use a much higher learning rate compared to non-LoRA fine-tuning. Here we use *1e-4* instead of the usual *1e-5*. Also, by using LoRA, it's possible to run `train_text_to_image_lora.py` in consumer GPUs like T4 or V100.**
+
+The final LoRA embedding weights have been uploaded to [sayakpaul/sd-model-finetuned-lora-t4](https://huggingface.co/sayakpaul/sd-model-finetuned-lora-t4). **___Note: [The final weights](https://huggingface.co/sayakpaul/sd-model-finetuned-lora-t4/blob/main/pytorch_lora_weights.bin) are only 3 MB in size, which is orders of magnitudes smaller than the original model.**
```
Comment (patrickvonplaten) [inline examples/text_to_image/README.md:167]: Very nice!

File: examples/text_to_image/train_text_to_image_lora.py (hunk 3 above)
Comment (patrickvonplaten) [reply examples/text_to_image/train_text_to_image_lora.py:2]: Don't think there is really an original repo we copied from - we just took the idea. Wouldn't change anything here.

File: examples/text_to_image/requirements.txt [hunk 15]
```diff
@@ -4,4 +4,6 @@ transformers>=4.25.1
 datasets
 ftfy
 tensorboard
+modelcards
```
Comment (patrickvonplaten) [inline examples/text_to_image/requirements.txt:7]:
```suggestion
```

Comment (patrickvonplaten) [reply examples/text_to_image/requirements.txt:7]: actually just using pure text here to create the model card

File: examples/text_to_image/train_text_to_image_lora.py [hunk 16]
```diff
@@ -0,0 +1,824 @@
...
+            hidden_size=hidden_size, cross_attention_dim=cross_attention_dim
+        )
+
+    unet.set_attn_processor(lora_attn_procs)
+    lora_layers = AttnProcsLayers(unet.attn_processors)
+
+    state_dict = lora_layers.state_dict()
+    lora_layers.load_state_dict(state_dict)
+
+    accelerator.register_for_checkpointing(lora_layers)
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply examples/text_to_image/train_text_to_image_lora.py:458]: It's in `accelerator.prepare` so we can remove it here

File: examples/text_to_image/train_text_to_image_lora.py [hunk 17]
```diff
@@ -0,0 +1,814 @@
...
+            hidden_size = unet.config.block_out_channels[block_id]
+
+        lora_attn_procs[name] = LoRACrossAttnProcessor(
+            hidden_size=hidden_size, cross_attention_dim=cross_attention_dim
+        )
+
+    unet.set_attn_processor(lora_attn_procs)
+    lora_layers = AttnProcsLayers(unet.attn_processors)
+
+    accelerator.register_for_checkpointing(lora_layers)
```
Comment (patrickvonplaten) [inline examples/text_to_image/train_text_to_image_lora.py:458]:
```suggestion
```

**Comment by patrickvonplaten** [comment]
Thanks a lot for your work here @sayakpaul - let's try to advertise it nicely now :-)
